File: src/networking/server/ServerNetworkHandler.java
package networking.server;

import engine.entities.Player;
import engine.model.enums.ObjectType;
import javafx.scene.transform.Rotate;
import networking.packets.*;
import server.ServerGameState;

public class ServerNetworkHandler {

    private ServerGameState gameState;
    private ConnectedPlayers connectedPlayers;
    private ServerNetworkDispatcher dispatcher;

    protected ServerNetworkHandler(ServerGameState gameState, ConnectedPlayers connectedPlayers, ServerNetworkDispatcher dispatcher) {
        this.gameState = gameState;
        this.connectedPlayers = connectedPlayers;
        this.dispatcher = dispatcher;
    }

	public void receiveHello(HelloPacket packet) {
		// TODO - integrate and get these values from somewhere
		int players = this.gameState.getAllPlayers().size();
		int maxPlayers = this.gameState.getNumPlayers();

		Packet response = new HelloAckPacket(players, maxPlayers, this.gameState.getGameType());
		this.dispatcher.send(response, packet.getIpAddress(), packet.getPort());
	}

	public void receiveConnect(ConnectPacket packet) {
		boolean isStarted = this.gameState.isStarted();
		boolean hasSpace = this.gameState.getAllPlayers().size() < this.gameState.getNumPlayers();
		System.out.println("Does the game have space: " + hasSpace);
		System.out.println("has the game Started: " + isStarted);

		// Allow connection if the game has not started yet and we have space for more
		// players
		ConnectAckPacket.Status status;
		int playerId = 0;
		if (isStarted) {
			status = ConnectAckPacket.Status.ERR_GAME_STARTED;
		} else if (!hasSpace) {
			status = ConnectAckPacket.Status.ERR_MAX_PLAYERS;
		} else {
			status = ConnectAckPacket.Status.SUC_CONNECTED;

			Player player = new Player(ObjectType.PLAYER);
			playerId = player.getId();

			this.connectedPlayers.addConnection(player, packet.getIpAddress(), packet.getPort());
			this.gameState.getAllPlayers().add(player);
			this.gameState.getObjects().add(player);

			System.out.println("Number of players connected: " + gameState.getAllPlayers().size());

			System.out.println("New player connection. P: " + playerId + " from: " + packet.getIpAddress());

		}

		Packet response = new ConnectAckPacket(playerId, status);
		this.dispatcher.send(response, packet.getIpAddress(), packet.getPort());

		if (status == ConnectAckPacket.Status.SUC_CONNECTED) {
            this.dispatcher.broadcastConnectedUser(playerId);
		}
	}

	public void receiveInitialGameStateAck(InitialGameStateAckPacket packet) {
		PlayerConnection playerConn = getPlayerConnection(packet);

		if (playerConn != null) {
			playerConn.setHasInitialState();
		} else {
			// Player connection not found
		}
	}

	public void receiveLocationState(LocationStatePacket packet) {
		PlayerConnection playerConn = getPlayerConnection(packet);

		if (playerConn != null) {
			Player player = playerConn.getPlayer();

			// Just update the location for now
			// TODO - validate if the location
			player.setLocation(packet.getX(), packet.getY());
			Rotate playerAngle = player.getPlayerAngle();
			playerAngle.setAngle(packet.getPlayerAngle());
			/*float playerCurrentHealth = packet.getPlayerHealth();
			player.setHealth(playerCurrentHealth); */
		} else {
			// Player connection not found
		}
	}

	public void receiveActionState(ActionStatePacket packet) {
		//currently set to broadcast if and only if there are currently more than 2 players
		boolean isStarted = this.gameState.isStarted();
		int numberOfPlayers = this.gameState.getAllPlayers().size();

	    PlayerConnection playerConn = getPlayerConnection(packet);
	    playerConn.getPlayer().doAction(packet.getAction());

	    if(isStarted && (numberOfPlayers >= 2) ){
            this.dispatcher.broadcastActionState(playerConn.getId(), packet.getAction());
	    }
	}

	public void receiveElementState(ElementStatePacket packet) {
		boolean isStarted = this.gameState.isStarted();
		int numberOfPlayers = this.gameState.getAllPlayers().size();

		PlayerConnection playerConn = getPlayerConnection(packet);
		playerConn.getPlayer().setCurrentElement(packet.getPlayerElementState());

		if(isStarted && (numberOfPlayers >= 2)) {
            this.dispatcher.broadcastElementState(playerConn.getId(), packet.getPlayerElementState());
	    }
	}

	/**
	 * Get the {@link PlayerConnection} for the sender of the Packet.
	 *
	 * @param packet The received packet.
	 * @return The {@link PlayerConnection} the packet was received from.
	 */
	private PlayerConnection getPlayerConnection(Packet packet) {
		return this.connectedPlayers.getPlayerConnection(packet.getIpAddress(), packet.getPort());
    }

}

File: neon-element-cpp/src/networking/server/ServerNetworkHandler.cpp
#include "ServerNetworkHandler.hpp"
#include "ServerNetworkDispatcher.hpp"
#include "networking/packets/HelloPackets.hpp"
#include "networking/packets/ConnectPackets.hpp"
#include "networking/packets/DisconnectPackets.hpp"
#include "networking/packets/StatePackets.hpp"
#include "networking/packets/GamePackets.hpp"
#include "engine/model/enums/ObjectType.hpp"
#include "engine/model/GameType.hpp"
#include "engine/model/gametypes/FirstToXKillsGame.hpp"
#include "engine/model/gametypes/TimedGame.hpp"
#include "engine/model/gametypes/HillGame.hpp"
#include "engine/model/gametypes/Regicide.hpp"

namespace neon::net {

// Helper: get the game param (kills needed, time, etc.) from the game type
static int getGameParam(const GameType& gt)
{
    switch (gt.getType()) {
        case GameType::Type::FirstToXKills:
            return static_cast<const FirstToXKillsGame&>(gt).getKillsNeeded();
        case GameType::Type::Timed:
            return static_cast<int>(static_cast<const TimedGame&>(gt).getDuration());
        case GameType::Type::Hill:
            return static_cast<const HillGame&>(gt).getScoreNeeded();
        case GameType::Type::Regicide:
            return static_cast<const Regicide&>(gt).getScoreNeeded();
    }
    return 0;
}

void ServerNetworkHandler::receiveHello(HelloPacket& packet)
{
    auto& sender = packet.getSender();
    HelloAckPacket ack(
        players_.count(),
        state_.getNumPlayers(),
        state_.getGameType().getType(),
        getGameParam(state_.getGameType()));
    dispatcher_.sendTo(ack, sender.address, sender.port);
}

void ServerNetworkHandler::receiveConnect(ConnectPacket& packet)
{
    auto& sender = packet.getSender();

    // Check if game already started
    if (state_.isStarted()) {
        ConnectAckPacket ack(0, ConnectAckPacket::Status::ErrGameStarted);
        dispatcher_.sendTo(ack, sender.address, sender.port);
        return;
    }

    // Check max players
    if (players_.count() >= state_.getNumPlayers()) {
        ConnectAckPacket ack(0, ConnectAckPacket::Status::ErrMaxPlayers);
        dispatcher_.sendTo(ack, sender.address, sender.port);
        return;
    }

    // Create new player and add to game
    auto player = std::make_unique<Player>(ObjectType::Player);
    int playerId = player->getId();
    Player* playerPtr = player.get();

    state_.getObjects().push_back(std::move(player));
    state_.rebuildPlayerList();

    players_.addConnection(playerPtr, sender.address, sender.port);

    // Acknowledge
    ConnectAckPacket ack(playerId, ConnectAckPacket::Status::Success);
    dispatcher_.sendTo(ack, sender.address, sender.port);

    // Broadcast to all
    dispatcher_.broadcastConnectedUser(playerId);
}

void ServerNetworkHandler::receiveDisconnect(DisconnectPacket& packet)
{
    auto& sender = packet.getSender();
    DisconnectAckPacket ack(true);
    dispatcher_.sendTo(ack, sender.address, sender.port);
}

void ServerNetworkHandler::receiveLocationState(LocationStatePacket& packet)
{
    auto& sender = packet.getSender();
    auto* conn = players_.getPlayerConnection(sender.address, sender.port);
    if (!conn) return;

    auto* player = conn->getPlayer();
    player->setLocation(static_cast<float>(packet.getX()),
                        static_cast<float>(packet.getY()));
    player->setAngle(static_cast<float>(packet.getAngle()));

    if (state_.isStarted() && players_.count() >= 2) {
        dispatcher_.broadcastLocationState(
            player->getId(), packet.getX(), packet.getY(), packet.getAngle());
    }
}

void ServerNetworkHandler::receiveActionState(ActionStatePacket& packet)
{
    auto& sender = packet.getSender();
    auto* conn = players_.getPlayerConnection(sender.address, sender.port);
    if (!conn) return;

    auto* player = conn->getPlayer();
    player->setCurrentAction(packet.getAction());

    if (state_.isStarted() && players_.count() >= 2) {
        dispatcher_.broadcastActionState(player->getId(), packet.getAction());
    }
}

void ServerNetworkHandler::receiveElementState(ElementStatePacket& packet)
{
    auto& sender = packet.getSender();
    auto* conn = players_.getPlayerConnection(sender.address, sender.port);
    if (!conn) return;

    auto* player = conn->getPlayer();
    player->setCurrentElement(packet.getElement());

    if (state_.isStarted() && players_.count() >= 2) {
        dispatcher_.broadcastElementState(player->getId(), packet.getElement());
    }
}

void ServerNetworkHandler::receiveReadyState(ReadyStatePacket& /*packet*/)
{
    // Ready state tracking for lobby — currently a stub
}

void ServerNetworkHandler::receivePowerUp(PowerUpPacket& /*packet*/)
{
    // Power-up pickup handled server-side via physics — stub
}

void ServerNetworkHandler::receiveInitialGameStateAck(InitialGameStateAckPacket& packet)
{
    auto& sender = packet.getSender();
    auto* conn = players_.getPlayerConnection(sender.address, sender.port);
    if (conn) {
        conn->setHasInitialState();
    }
}

} // namespace neon::net
